scio-core/src/main/scala/com/spotify/scio/ScioContext.scala

```diff
@@ -614,7 +618,8 @@ class ScioContext private[scio] (val options: PipelineOptions,
    */
   def pubsubTopic[T: ClassTag](topic: String,
                                idLabel: String = null,
-                               timestampLabel: String = null)
+                               timestampLabel: String = null,
+                               parseFn: gio.PubsubIO.PubsubMessage => T = null)
```

Comment:
How would this work with types like Avro? In that case `T` is an Avro `SpecificRecord` and may or may not contain fields that can represent the extra attributes? Or the user has to decode manually from `byte[]` to Avro?

Looking at `PubsubMessage` it's just 2 fields, `byte[] message` and `java.util.Map<java.lang.String,java.lang.String> attributes`.

How about variations of these 2 methods that return `[(T, Map[String, String])]` where `T` is the main message and `Map` is the attributes? That way we can still use the inferred coder. And the API feels more idiomatic scala.